src/identity/interaction/util/InteractionCompleter.ts

```diff
@@ -1,27 +1,34 @@
+import { ServerResponse } from 'http';
 import type { InteractionResults } from 'oidc-provider';
-import type { HttpHandlerInput } from '../../../server/HttpHandler';
+import type { HttpRequest } from '../../../server/HttpRequest';
 import { AsyncHandler } from '../../../util/handlers/AsyncHandler';
 import type { ProviderFactory } from '../../configuration/ProviderFactory';
 
+/**
+ * Parameters required to specify how the interaction should be completed.
+ */
 export interface InteractionCompleterParams {
   webId: string;
   shouldRemember?: boolean;
 }
 
-export type InteractionCompleterInput = HttpHandlerInput & InteractionCompleterParams;
+export interface InteractionCompleterInput extends InteractionCompleterParams {
+  request: HttpRequest;
+}
 
 /**
  * Completes an IDP interaction, logging the user in.
+ * Returns the URL the request should be redirected to.
  */
-export class InteractionCompleter extends AsyncHandler<InteractionCompleterInput> {
+export class InteractionCompleter extends AsyncHandler<InteractionCompleterInput, string> {
```

Comment:
I wonder if we want `UrlString` or so.

Reply:
We could do that. I'm guessing we probably have several places where we can make good use of template literal types.

Reply:
Template literals might be hard; typing a URL is non-trivial.